File: 백준/Gold/20033. Square， Not Rectangle/Square， Not Rectangle.py
import sys
input = lambda: sys.stdin.readline().strip()

n = int(input())
arr = list(map(int, input().split()))


start = 1
end = 10 ** 9
answer = 0

while start <= end:
    mid = (start + end) // 2


    if n < mid:
        end = mid - 1
    flag = False
    count = 0
    result = 0
    for i in arr:
        if not flag:
            if i - mid >= 0:
                count += 1
                flag = True

        else:
            if i - mid >= 0:
                count += 1
            else:
                result = max(result, count)
                count = 0
                flag = False
    result = max(result, count)


    if result >= mid:
        answer = max(answer, mid)
        start = mid + 1
    else:
        end = mid - 1

print(answer)

File: 백준/Gold/20033. Square， Not Rectangle/Square， Not Rectangle.java
import java.io.*;
import java.util.*;

public class Main {
	static int N;
	static int[] arr;

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = null;
		N = Integer.parseInt(br.readLine());
		arr = new int[N];
		st = new StringTokenizer(br.readLine());

		for (int i = 0; i < N; i++) {
			arr[i] = Integer.parseInt(st.nextToken());
		}
		System.out.println(solution());
	}

	static int solution() {
		int start = 1;
		int end = N;
		int res = 0;

		while (start <= end) {
			int mid = (start + end) / 2;
			int cnt = 0;
			boolean isBreak = false;
			for (int i = 0; i < N; i++) {
				if (arr[i] >= mid) {
					if (++cnt >= mid) {
						isBreak = true;
						break;
					}
				} else {
					cnt = 0;
				}
			}
			if (isBreak) {
				res = Math.max(res, mid);
				start = mid + 1;
			} else {
				end = mid - 1;
			}
		}
		return res;
	}
}